internal/pkg/pluginengine/topological_sort.go

```diff
@@ -0,0 +1,82 @@
+package pluginengine
+
+import (
+	"fmt"
+
+	"github.com/merico-dev/stream/internal/pkg/configloader"
+	"github.com/merico-dev/stream/pkg/util/log"
+)
+
+func generateKeyFromTool(tool configloader.Tool) string {
+	return fmt.Sprintf("%s.%s", tool.Name, tool.Plugin.Kind)
+}
+
+func topologicalSort(tools []configloader.Tool) ([][]configloader.Tool, error) {
+	// the final result that contains sorted Tools
+	// it's a sorted/ordered slice,
+	// each element is a slice of Tools that can run parallel without any particular order
+	res := make([][]configloader.Tool, 0)
+
+	// a "graph", which contains "nodes" that haven't been processed yet
+	unprocessedNodeSet := make(map[string]bool)
+	for _, tool := range tools {
+		key := generateKeyFromTool(tool)
+		unprocessedNodeSet[key] = true
+	}
+
+	// while there is still a node in the graph left to be processed:
+	for len(unprocessedNodeSet) > 0 {
+		// the next batch of tools that can run in parallel
+		batch := make([]configloader.Tool, 0)
+
+		for _, tool := range tools {
+			// if the tool has already been processed (not in the unprocessedNodeSet anymore), pass
+			key := generateKeyFromTool(tool)
+			if _, ok := unprocessedNodeSet[key]; !ok {
+				continue
+			}
+
+			// if there isn't any dependency: it's the "start" of the graph
+			// we can put it into the first batch
+			if len(tool.DependsOn) == 0 {
+				log.Debugf("TOOL %s.%s dependency already solved\n", tool.Name, tool.Plugin.Kind)
+				batch = append(batch, tool)
+			} else {
+				dependency_resolved := true
+				for _, dep := range tool.DependsOn {
```

Comment:
Too many nesting levels,  should a new function be abstracted?

Reply:
Very good point!